### ISSUE TITLE
case when _ matching only works on literal tuples

### ISSUE THREAD

**Issue by postmodern** [crystal-lang/crystal#8818]
It's not clear why `_` matching within case/when statements only works on literal tuples, and not tuple variables. This behavior seems inconsistent to me.

### With literal tuple

```
case {1,2}
when {1, _}
  puts "starts with 1"
else
  puts "nope"
end
```

### With tuple variable

var = {1,2}

case var
when {1,_}
  puts "starts with 1"
else
  puts "nope"
end
```

### Expected result

Both compile.

### Actual result

The tuple variable example fails to compile:

```
 4 | when {1, _}
              ^
Error: can't read from _
```

**Comment by asterite** [comment]
Yes, tuple. matching like that is done specifically for. matching explicit tuple literals with the same number of arguments.

This is a limitation but it's on purpose. There's no pattern matching in Crystal, just this explicit tuple matching.

**Comment by straight-shoota** [comment]
The error message seems a bit cryptic. Could be improved.

**Comment by wontruefree** [comment]
@asterite what is the reason for the limitation?

**Comment by asterite** [comment]
The reason is that:

```
case var
when {something, something_else}
```

would have to be translated to something like:

```crystal
if var.is_a?(Tuple) && "tuple size is 2" && ...
```

and it looks more like pattern matching than anything else. And if we do that then people will start asking "well, why not do the same but allow array literals? or hash literals?".

I just didn't want to introduce a partial pattern matching.

So right now you can just match an explicit tuple literal with other explicit matches, which actually doesn't involve tuples at all: it's just doing matching on multiple values at the same time.

**Comment by asterite** [comment]
In fact I regret the case tuple-matching "feature" and I'd like to remove it if we still have time.

It makes the language a bit more complex, it's a bit harder to do exhaustiveness check, it makes this kind of questions spawn, etc.

**Comment by Sija** [comment]
@asterite It would be better to work on bringing the pattern matching as a 1st class citizen to the language.

**Comment by wontruefree** [comment]
I could see this feature being really confusing from a refactor standpoint too.  I think it would be worth improving the error message at least and possibly removing it.

**Comment by straight-shoota** [comment]
IMO it's still a useful feature and since the limitations are clear, there's not much confusion around it.
It wouldn't be a huge loss, if it was removed, though.

**Comment by postmodern** [comment]
@straight-shoota could you give an example of a useful scenario? What struck me as odd is why would you need something like `_`, when you are required to specify the tuple literally. At that point, I'd think an `if/elsif` statement would be simpler, since you already have the elements which make up the tuple in scope.

I kind of agree this is an all-or-nothing feature. Pseudo pattern-matching would confuse new users and create similar issues such as this one.

Improving the error message or updating the docs might help you reframe the feature as not-quite-pattern-matching to avoid future confusion.

**Comment by straight-shoota** [comment]
The main advantage is that the code is more concise. Arguments are always at the same position and it's easy to relate to the meaning of each branch.
```cr
case {a, b}
when {foo, _}
when {boo, bar}
when {boo, baz}
end
# vs.
if a == foo
elsif a == boo && b == bar
elsif a == boo && b == baz
end
# vs.
if a == foo
elsif a == boo
  if b == bar
  elsif b == baz
  end
end
```
